pip/src/interop.rs

```diff
@@ -0,0 +1,657 @@
+    )
+    .map_err(into_estimation_errors)?;
+
+    resource_estimator::estimate_entry(&mut interpreter, params)
+}
+
+/// Converts a list of Q# errors into a list of resource estimator errors.
+fn into_estimation_errors(errors: Vec<interpret::Error>) -> Vec<resource_estimator::Error> {
+    errors
+        .into_iter()
+        .map(|error| resource_estimator::Error::Interpreter(error.clone()))
+        .collect::<Vec<_>>()
+}
+
+/// Formats a list of QASM3 errors into a single string.
+pub(crate) fn format_qasm_errors(errors: Vec<WithSource<qsc_qasm3::Error>>) -> String {
+    errors
+        .into_iter()
+        .map(|e| {
+            let mut message = String::new();
+            let report = miette::Report::new(e);
+            write!(message, "{report:?}").unwrap();
+            message
+        })
+        .collect::<String>()
+}
+
+/// Creates a `FileSystem` from the provided Python callbacks.
+/// If any of the callbacks are missing, this will panic.
+pub(crate) fn create_filesystem_from_py(
+    py: Python,
+    read_file: Option<PyObject>,
+    list_directory: Option<PyObject>,
+    resolve_path: Option<PyObject>,
+    fetch_github: Option<PyObject>,
+) -> impl FileSystem + '_ {
+    file_system(
+        py,
+        read_file.expect("file system hooks should have been passed in with a read file callback"),
+        list_directory
+            .expect("file system hooks should have been passed in with a list directory callback"),
+        resolve_path
+            .expect("file system hooks should have been passed in with a resolve path callback"),
+        fetch_github
+            .expect("file system hooks should have been passed in with a fetch github callback"),
+    )
+}
+
+/// Creates an `Interpreter` from the provided AST package and configuration.
+fn create_interpreter_from_ast(
+    ast_package: Package,
+    source_map: SourceMap,
+    profile: Profile,
+    language_features: LanguageFeatures,
+    package_type: PackageType,
+) -> Result<Interpreter, Vec<interpret::Error>> {
+    let mut store = PackageStore::new(qsc::compile::core());
+    let mut dependencies = Vec::new();
+
+    let capabilities = profile.into();
+
+    dependencies.push((store.insert(qsc::compile::std(&store, capabilities)), None));
+    let (mut unit, errors) = qsc::compile::compile_ast(
+        &store,
+        &dependencies,
+        ast_package,
+        source_map,
+        package_type,
+        capabilities,
+    );
+
+    if !errors.is_empty() {
+        return Err(into_errors(errors));
+    }
+
+    unit.expose();
+    let source_package_id = store.insert(unit);
+
+    interpret::Interpreter::from(
+        store,
+        source_package_id,
+        capabilities,
+        language_features,
+        &dependencies,
+    )
+}
+
+/// Sanitizes the name to ensure it is a valid identifier according
+/// to the Q# specification. If the name is empty, returns "circuit".
+pub(crate) fn sanitize_name<S: AsRef<str>>(name: S) -> String {
```

Comment:
should we reuse this logic when we transform file names into namespace names for implicit namespaces?